agent/php_mysqli.c

```diff
@@ -398,39 +398,54 @@ static nr_status_t nr_php_mysqli_link_real_connect(
     zval* link,
     const nr_mysqli_metadata_link_t* metadata TSRMLS_DC) {
   zend_ulong argc = 0;
+  zend_ulong arg_required = 0;
   zval* argv[7] = {0};
   zend_ulong i;
   zval* retval = NULL;
 
-#define ADD_IF_INT_SET(args, argc, value) \
-  if (value) {                            \
-    args[argc] = nr_php_zval_alloc();     \
-    ZVAL_LONG(args[argc], value);         \
-    argc++;                               \
-  }
-
-#define ADD_IF_STR_SET(args, argc, value) \
-  if (value) {                            \
-    args[argc] = nr_php_zval_alloc();     \
-    nr_php_zval_str(args[argc], value);   \
-    argc++;                               \
-  }
-
-  ADD_IF_STR_SET(argv, argc,
+#define ADD_IF_INT_SET(is_set, args, argc, value) \
+  if (value) {                                    \
+    args[argc] = nr_php_zval_alloc();             \
+    ZVAL_LONG(args[argc], value);                 \
+    argc++;                                       \
+  } else if (false == is_set) {                   \
+    args[argc] = nr_php_zval_alloc();             \
+    ZVAL_NULL(args[argc]);                        \
+    argc++;                                       \
+  }
+
+#define ADD_IF_STR_SET(is_set, args, argc, value) \
+  if (value) {                                    \
+    args[argc] = nr_php_zval_alloc();             \
+    nr_php_zval_str(args[argc], value);           \
+    argc++;                                       \
+  } else if (false == is_set) {                   \
+    args[argc] = nr_php_zval_alloc();             \
+    ZVAL_NULL(args[argc]);                        \
+    argc++;                                       \
+  }
+
+  ADD_IF_STR_SET(true, argv, argc,
                  nr_php_mysqli_strip_persistent_prefix(metadata->host));
-  ADD_IF_STR_SET(argv, argc, metadata->user);
-  ADD_IF_STR_SET(argv, argc, metadata->password);
-
+  ADD_IF_STR_SET(true, argv, argc, metadata->user);
+  ADD_IF_STR_SET(true, argv, argc, metadata->password);
   /*
    * We can only add the remaining metadata fields if we already have three
    * arguments (host, user and password) above, lest we accidentally set the
    * wrong positional argument to something it doesn't mean.
    */
+  arg_required = argc;
   if (argc == 3) {
-    ADD_IF_STR_SET(argv, argc, metadata->database);
-    ADD_IF_INT_SET(argv, argc, metadata->port);
-    ADD_IF_STR_SET(argv, argc, metadata->socket);
-    ADD_IF_INT_SET(argv, argc, metadata->flags);
+    ADD_IF_STR_SET(false, argv, argc, metadata->database);
+    ADD_IF_INT_SET(false, argv, argc, metadata->port);
+    ADD_IF_STR_SET(false, argv, argc, metadata->socket);
+
+    /*
+     * Avoid defaulting the final $flags argument to null.
+     */
+    if (metadata->flags) {
```

Comment:
Do we still need this check if we can use is_set/default_null to accomplish the same check?

Reply:
Good point, removed the check here: [b1c7ab5](https://github.com/newrelic/newrelic-php-agent/pull/976/commits/b1c7ab503d97f9c0b67dba9d5a92ed36a8c35902)

Reply:
:+1: